pynxtools/dataconverter/readers/multi/reader.py

```diff
@@ -0,0 +1,208 @@
+    callbacks: Optional[ParseJsonCallbacks] = None,
+) -> dict:
+    """
+    Parses a json file and returns the data as a dictionary.
+    """
+    if callbacks is None:
+        # Use default callbacks if none are explicitly provided
+        callbacks = ParseJsonCallbacks()
+
+    config_file_dict = parse_flatten_json(file_path)
+    fill_wildcard_data_indices(config_file_dict, callbacks.dims)
+
+    for key, value in config_file_dict.items():
+        if not isinstance(value, str) or ":" not in value:
+            continue
+
+        precursor = value.split(":")[0]
+        value = value[value.index(":") + 1 :]
+
+        config_file_dict[key] = callbacks.apply_special_key(precursor, key, value)
+
+    return config_file_dict
+
+
+class MultiFormatReader(BaseReader):
+    """
+    A reader that takes a mapping json file and a data file/object to return a template.
+    """
+
+    # Whitelist for the NXDLs that the reader supports and can process
+    supported_nxdls: List[str] = []
+    extensions: Dict[str, Callable[[Any], dict]] = {
+        ".yaml": parse_yml,
+        ".yml": parse_yml,
+    }
+    kwargs: Optional[Dict[str, Any]] = None
+
+    def __init__(self):
+        self.extensions[".json"] = lambda fn: parse_json_config(
+            fn,
+            ParseJsonCallbacks(
+                attrs_callback=self.get_attr,
+                data_callback=self.get_data,
+                dims=self.get_data_dims,
+            ),
+        )
+
+    def setup_template(self) -> Dict[str, Any]:
+        """
+        Setups the initial data in the template.
+        This may be used to set fixed information, e.g., about the reader.
+        """
+        return {}
+
+    # pylint: disable=unused-argument
+    def handle_objects(self, objects: Tuple[Any]) -> Dict[str, Any]:
+        """
+        Handles the objects passed into the reader.
+        """
+        return {}
+
+    def get_attr(self, path: str) -> Any:
+        """
+        Returns an attributes from the given path.
+        """
+        return {}
+
+    def get_data(self, path: str) -> Any:
+        """
+        Returns data from the given path.
+        """
+        return {}
+
+    def get_data_dims(self, path: str) -> List[str]:
+        """
+        Returns the dimensions of the data from the given path.
+        """
+        return []
+
+    def read(
+        self,
+        template: dict = None,
+        file_paths: Tuple[str] = None,
+        objects: Tuple[Any] = None,
+        **kwargs,
+    ) -> dict:
+        """
+        Reads data from multiple files and passes them to the appropriate functions
+        in the extensions dict.
+        """
+        template = Template()
+        self.kwargs = kwargs
+
+        def sort_keys(filename: str) -> str:
+            """
+            Makes sure to read json and yaml files last
+            """
+            ending = os.path.splitext(filename)[1]
+            return ".zzzzz" if ending in (".json", ".yaml", ".yml") else ending
```

Comment:
This effectively means the ELN data always get priority over the actual files, even if you have the `@attrs` keyword, right?

Reply:
Not necessarily, they would have the same sort key so it depends on their initial order (so it's kind of arbitrary from user input). I also thought about separating this and always do the config replacements last.